src/main/rx/cc2500_redpine.c

```diff
@@ -0,0 +1,194 @@
+    rcData[14]= (packet[CHANNEL_START + 6] & 0x40) ? PWM_RANGE_MAX : PWM_RANGE_MIN; 
+    rcData[15]= (packet[CHANNEL_START + 6] & 0x80) ? PWM_RANGE_MAX : PWM_RANGE_MIN;         
+}
+
+rx_spi_received_e redpineHandlePacket(uint8_t * const packet, uint8_t * const protocolState)
+{
+    static uint16_t looptime = TOTAL_PACKET_TIME;
+    static timeUs_t packetTimerUs;
+
+    static timeUs_t totalTimerUs;
+ 
+    rx_spi_received_e ret = RX_SPI_RECEIVED_NONE;
+
+    switch (*protocolState) {
+    case STATE_STARTING:
+        listLength = 47;
+        initialiseData(0);
+        *protocolState = STATE_UPDATE;
+        nextChannel(1);
+        cc2500Strobe(CC2500_SRX);
+#ifdef USE_RX_FRSKY_SPI_PA_LNA
+        cc2500TxDisable();
+#endif // USE_RX_FRSKY_SPI_PA_LNA
+
+        break;
+    case STATE_UPDATE:
+        packetTimerUs = 0;
+        totalTimerUs = micros();
+        *protocolState = STATE_DATA;
+        if (rxSpiCheckBindRequested(false)) {
+            packetTimerUs = 0;
+            missingPackets = 0;
+            *protocolState = STATE_INIT;
+            break;
+        }
+
+        FALLTHROUGH;
+        // here FS code could be
+    case STATE_DATA:
+        if (cc2500getGdo()) {
+            uint8_t ccLen = cc2500ReadReg(CC2500_3B_RXBYTES | CC2500_READ_BURST) & 0x7F;
+            DEBUG_SET(DEBUG_RX_FRSKY_SPI, 3, ccLen);
+            if (ccLen > RX_SPI_MAX_PAYLOAD_SIZE) {
+                 cc2500Strobe(CC2500_SFRX);
+            } else if (ccLen) {
+                cc2500ReadFifo(packet, ccLen);
+                uint16_t lcrc= calculateCrc(&packet[0], CHANNEL_START+9);
+
+                if(((lcrc >> 8) == packet[CHANNEL_START+9]) && 
+                    ((lcrc&0x00FF) == packet[CHANNEL_START+10]) &&
+                    (packet[1] == rxFrSkySpiConfig()->bindTxId[0]) &&
+                    (packet[2] == rxFrSkySpiConfig()->bindTxId[1])) {
+                    
+                    looptime = packet[CHANNEL_START+7] * 100;
+
+                    DEBUG_SET(DEBUG_RX_FRSKY_SPI, 0, looptime);
+                    DEBUG_SET(DEBUG_RX_FRSKY_SPI, 1, packet[CHANNEL_START+8] * 10);
+                    packetTimerUs = micros();
+                    totalTimerUs = micros();                         
+
+                    missingPackets = 0;
+                    rxSpiLedOn();
+
+                    cc2500setRssiDbm(packet[ccLen - 2]);
+
+                    ret = RX_SPI_RECEIVED_DATA;
+                    nextChannel(1);
+                    cc2500Strobe(CC2500_SRX);
+                } 
+            }
+        }
+
+        if (cmpTimeUs(micros(), totalTimerUs) > 50 * looptime) {
+            //out of sync with packets - do a complete resysnc
+            rxSpiLedToggle();
+            setRssiDirect(0, RSSI_SOURCE_RX_PROTOCOL);
+            nextChannel(1);
+            cc2500Strobe(CC2500_SRX);
+            *protocolState = STATE_UPDATE;
+        } else if ((cmpTimeUs(micros(), packetTimerUs) > looptime) && packetTimerUs) {
```

Comment:
Clearly if you miss a packet you need to switch the channel before the tx starts sending again so you have a chance to receive the packet. It's not clear from this section of code which looptime the tx sends -I'll assume normal time between packets. In that case it would be better to change the channel at the earliest time you know that you didn't receive the last one. That'll be way earlier than packetTimerUs + looptime.

Reply:
At some point ideally we'll switch channels and record receive rx timestamps using irqs or even DMA. This will get all the more important the more longer running tasks we have.

Reply:
Yeah looptime is the time between the start between two TX packets. The looptime getting set is sort of a debug tool right now, but works pretty well.  With a menu option in deviation I can adjust the time on the transmitter see how fast I can go before it breaks or extend it out to 25ms to see if I can feel a difference.  

I was doing some work to try and detect the time and keep a running average, but I didn't have it working consistently.  That might be closer to allow differences between tx and rx.

There is also probably jitter to contend with as well on the TX side (RX as well i suppose).  So it seems like you'd want to your switch as soon as you expect a packet to be done + a little margin.  Just in case the packet is taking a little longer to finish getting received.

Actually maybe the best would be to switch half way between the expected end of a packet and the start of a new one?  That would give a extra buffer room for any jitter or mismatch between the tx and rx.